File: adventOfCode/Python/Day15.py
#Advent of Code: Day 15
import time

#compute the next generation. 
#	If picky, it will search for a number which is evenly dividable by modulo
def cycle(generator, multiplier, divider, modulo, picky):
	generator = (generator * multiplier) % divider
	if picky:
		while (generator% modulo != 0):
			generator = (generator * multiplier) % divider
	return generator

#compare last 16 bits of both generations
def compareBinary(genA, genB):
	return (genA & 0xffff) == (genB & 0xffff)

#run part1 or part2
def part(genA, genB, multiA, multiB, divider, cycles, modA=0, modB=0, picky=False):
	count = 0
	for _ in range(cycles):
		if (compareBinary(genA, genB)):
			count += 1
		genA = cycle(genA, multiA, divider, modA, picky)
		genB = cycle(genB, multiB, divider, modB, picky)	
	return count
	
def main():
	#test input
	genA, genB = 65, 8921
	#own input
	genA, genB = 116, 299
	
	#const values
	multiA, multiB = 16807, 48271
	divider = 2147483647
	modA, modB = 4, 8

	# part 1 without being picky and doing 40M cycles
	cycles = 40000000
	count = part(genA, genB, multiA, multiB, divider, cycles)
	print "Part 1: found {} matches".format(count)
	#runs isolated (without part 2) in about 48 seconds
	
	# part 2, being picky and doing 5M cycles
	cycles = 5000000
	count = part(genA, genB, multiA, multiB, divider, cycles, modA, modB, True)
	print "Part 2: found {} matches".format(count)
	#runs isolated (without part 1) in about 33 seconds

if __name__ == "__main__":
	start_time = time.time()
	main()
	print("- %s seconds -" % (time.time() - start_time))

File: adventOfCode/Java/Day15.java

public class Day15 {

    public static void main(String[] args)  {
    	long startTime = System.nanoTime();      
    	
    	long genA = 116;
        long genB = 299;
        //System.out.println("part 1 " + generator(genA, genB, 40000000, false)); 
        System.out.println("part 2 " + generator(genA, genB, 5000000, true));

        long difference = System.nanoTime() - startTime;
        System.out.println("Elapsed time in seconds: " + difference / 1000000000.0);

    }       

    

    public static long getNextGenerator(long generator, boolean picky, int modulo, int multiplier) {
        generator = (generator * multiplier) % 2147483647;
        if (picky) {
        	while(generator % modulo != 0) {
                generator = (generator * multiplier) % 2147483647;
        	}
        }
        return generator;
    }
    
    public static int generator(long genA, long genB, int cycles, boolean picky) {
        int count = 0;
        for (int i = 0; i < cycles; i++) {
        	genA = getNextGenerator(genA, picky, 4, 16807);
            genB = getNextGenerator(genB, picky, 8, 48271);
            if ((genA & 65535) == (genB & 65535))
                count++;
        }
        return count;    	
    }

}